pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/operator/AggregateOperator.java

```diff
@@ -116,6 +119,11 @@ public String toExplainString() {
   protected TransferableBlock getNextBlock() {
     try {
       consumeInputBlocks();
+
+      if (!_readyToConstruct) {
```

Comment:
I was wondering what would be the best way to model the "stop-the-world" and "passthrough" operators?

my thinking is: model all operators as "take 0 - 1 block in, produce 0 - 1 block out"

e.g. if a chain calling `nextBlock()` 
- when it hits agg operator when it hasn't seen an EOS, then it is a 1 block in, 0 block out operator
- when it hits agg operator that has seen an EOS, then it is a 0 block in, 1 block out.
- when it hits filter operator, it is a 1 block in, 0 - 1 block out 
- when it hits project operator, it is a 1 block in, 1 block out 

with this design each operator is clearly a state machine and we don't have to maintain special boolean flags on each impl, what do you think?

(0 block out as No-op block out); 
(0 block in as not calling getNextBlock of the upstream operator)

Reply:
I like this idea a lot, I'll play around with it today

Reply:
I played around with this and couldn't easily come up with something that I liked significantly better than just inlining the logic. 

The reason for this was because:
1. some stateful operators have different logic (e.g. HashJoin needs to build the broadcast table first, pulling from the right side, and then becomes a 1:1 operator)
2. the state transition couldn't happen in the parent class, the child class still had to track when to transition itself, which means we still held onto state in the child class
3. it becomes harder to read when the state transition logic is split across two classes

I think for this PR, I'll keep the classes as is and we can try to refactor and make it easier to read in a separate PR.

Reply:
+1 no problem